urlc175f05cb8_mchllbrt_PermCode_tgz-pJ8-permlib_examples_UniversalSeparableJ8/src/permlab/examples/TableauState.java

```diff
@@ -0,0 +1,523 @@
+            for (int i = 0; i < row.size(); i++) {
+                result[i]++;
+            }
+        }
+        return result;
+    }
+
+    private static ArrayList<int[]> LSLeaves(int[] p) {
+        ArrayList<int[]> result = new ArrayList<int[]>();
+        LSProcess(Arrays.copyOf(p, p.length), result);
+        return result;
+    }
+
+    private static void LSProcess(int[] p, ArrayList<int[]> result) {
+        ArrayDeque<int[]> q = new ArrayDeque<int[]>();
+        q.add(p);
+        while (!q.isEmpty()) {
+            int[] v = q.poll();
+            if (avoids2143(v)) {
+                result.add(v);
+                continue;
+            }
+            q.addAll(children(v));
+        }
+
+
+    }
+
+    public static ArrayDeque<int[]> children(int[] elements) {
+        arrayChildren.clear();
+
+        RS rs = rs(elements);
+        int deltaS = 0;
+        for (int j = rs.r - 1; j >= 0; j--) {
+            if (elements[j] < elements[rs.s]) {
+                deltaS++;
+            } else {
+                deltaS--;
+            }
+            int deltaJ = 0;
+            for (int t = j + 1; t < rs.r; t++) {
+                if (elements[t] > elements[j]) {
+                    deltaJ++;
+                } else {
+                    deltaJ--;
+                }
+            }
+            if (deltaS + deltaJ == 1) {
+                int[] newElements = Arrays.copyOf(elements, elements.length);
+                int v = newElements[j];
+                newElements[j] = newElements[rs.s];
+                newElements[rs.s] = newElements[rs.r];
+                newElements[rs.r] = v;
+                arrayChildren.add(newElements);
+            }
+        }
+        if (arrayChildren.isEmpty()) {
+            int[] newElements = new int[elements.length + 1];
+            newElements[0] = 0;
+            for (int i = 1; i < newElements.length; i++) {
+                newElements[i] = elements[i - 1] + 1;
+            }
+            return children(newElements);
+        }
+        return arrayChildren;
+    }
+
+    public static RS rs(int[] elements) {
+        RS result = new RS();
+        int r = elements.length - 2;
+        while (r >= 0 && elements[r] < elements[r + 1]) {
+            r--;
+        }
+        result.r = r;
+        int s = elements.length - 1;
+        while (elements[s] > elements[r]) {
+            s--;
+        }
+        result.s = s;
+        return result;
+
+
+    }
+
+    static class RS {
+
+        int r, s;
+    }
+    static final int MAX_LENGTH = 100;
+    static final int[] max3s = new int[MAX_LENGTH];
+    static final int[] largeValues = new int[MAX_LENGTH];
+    static final int[] smallValues = new int[MAX_LENGTH];
+    static final ArrayDeque<int[]> aq = new ArrayDeque<int[]>();
+    static final ArrayDeque<int[]> arrayChildren = new ArrayDeque<int[]>();
+    static int largeStackTop = -1;
+    static int smallStackTop = -1;
+
+    public static boolean avoids2143(int[] values) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 398-450 into a separate method.